featurebyte/service/column_attributes.py

```diff
@@ -0,0 +1,157 @@
+    """
+
+    def __init__(
+        self,
+        preview_service: PreviewService,
+        feature_store_service: FeatureStoreService,
+    ):
+        self.preview_service = preview_service
+        self.feature_store_service = feature_store_service
+
+        self.detectors: list[BaseColumnAttributesDetector] = [
+            ArrayEmbeddingColumnAttributesDetector(),
+            FlatDictColumnAttributesDetector(),
+        ]
+
+    async def add_columns_attributes(self, table: TableModel) -> None:
+        """Detect and adds columns attributes like embedding, flat dict, etc.
+
+        Parameters
+        ----------
+        table : TableModel
+            Table with columns to run detection for.
+        """
+        feature_store = await self.feature_store_service.get_document(
+            document_id=table.tabular_source.feature_store_id,
+        )
+        source_table = SourceTable(
+            feature_store=feature_store,
+            tabular_source=table.tabular_source,
+            columns_info=table.columns_info,
+        )
+        graph, node = source_table.frame.extract_pruned_graph_and_node()
+        sample = dataframe_from_json(
+            await self.preview_service.sample(
+                FeatureStoreSample(
+                    graph=graph,
+                    node_name=node.name,
+                ),
+                size=SAMPLE_SIZE,
+                seed=RANDOM_SEED,
+            )
+        )
+        if sample.shape[0] > 0:  # pylint: disable=no-member
+            for detector in self.detectors:
+                await detector.detect(table.columns_info, sample)
+
+
+class BaseColumnAttributesDetector(ABC):
+    """Base column attributes detector interface."""
+
+    @abstractmethod
+    async def detect(self, columns_info: list[ColumnInfo], sample: pd.DataFrame) -> None:
+        """Given the sample of data, check if column of a concrete type has some specific attributes.
+        Updates ColumnInfo.attributes list with new attribute.
+
+        Examples of checks: array if the same dimension, flat dictionary, etc.
+
+        Parameters
+        ----------
+        columns_info : list[ColumnInfo]
+            List of columns to check
+        sample : pd.DataFrame
+            Data sample
+        """
+
+
+class ArrayEmbeddingColumnAttributesDetector(BaseColumnAttributesDetector):
+    """Detect and add attributes to columns which can be considered embeddings:
+    - 1 dimensional
+    - Same number of dimensions
+    - All values numeric and finite
+    """
+
+    async def detect(self, columns_info: list[ColumnInfo], sample: pd.DataFrame) -> None:
+        for column in columns_info:
+            if str(column.dtype) == DBVarType.ARRAY.value:
+                series = sample[column.name]
+                series = series[pd.notnull(series)]
+                shapes = series.apply(len)
+
+                if shapes.unique().ravel().shape != (1,):
+                    continue
```

Comment:
Would be good to add a comment here with we continue if the if statement is true.